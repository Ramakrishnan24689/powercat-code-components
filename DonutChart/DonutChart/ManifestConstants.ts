export const enum ManifestPropertyNames {
    dataset = 'dataset',
}

export const enum ItemColumns {
    Legend = 'ItemLegend',
    Key = 'ItemKey',
    Value = 'ItemValue',
    Color = 'ItemColor',
}
