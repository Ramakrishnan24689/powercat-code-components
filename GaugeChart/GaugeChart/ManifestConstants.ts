import { GaugeValueFormat } from '@fluentui/react-charting';

export const enum ManifestPropertyNames {
    dataset = 'dataset',
}

export const ChartValueFormats = {
    '1': GaugeValueFormat.Fraction,
    '0': GaugeValueFormat.Percentage,
};

export const enum ItemColumns {
    Legend = 'ItemLegend',
    Key = 'ItemKey',
    Count = 'ItemCount',
    Color = 'ItemColor',
    Size = 'ItemSize',
}
