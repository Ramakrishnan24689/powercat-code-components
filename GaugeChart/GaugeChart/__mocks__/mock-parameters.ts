/* istanbul ignore file */

import { IInputs } from '../generated/ManifestTypes';
import { MockEnumProperty, MockStringProperty, MockTwoOptionsProperty, MockWholeNumberProperty } from './mock-context';
import { MockDataSet } from './mock-datasets';

export function getMockParameters(): IInputs {
    return {
        Title: new MockStringProperty(),
        Sublabel: new MockStringProperty(),
        ChartValue: new MockWholeNumberProperty(),
        ChartValueFormat: new MockEnumProperty(),
        MinValue: new MockWholeNumberProperty(),
        MaxValue: new MockWholeNumberProperty(),
        HideLegend: new MockTwoOptionsProperty(),
        HideMinMax: new MockTwoOptionsProperty(),
        Theme: new MockStringProperty(),
        CustomColors: new MockTwoOptionsProperty(),
        items: new MockDataSet([]),
        TabIndex: new MockWholeNumberProperty(),
        Tooltip: new MockStringProperty(),
    };
}
