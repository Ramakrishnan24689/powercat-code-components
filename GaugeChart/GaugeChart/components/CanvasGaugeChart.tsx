import * as React from 'react';
import { GaugeChart, IGaugeChartProps } from '@fluentui/react-charting';
import { IPartialTheme, ThemeProvider, createTheme } from '@fluentui/react';

export type IGaugeChartCanvasProps = IGaugeChartProps & { themeJSON: string };

export const CanvasGaugeChart = React.memo((props: IGaugeChartCanvasProps) => {
    const { themeJSON, ...rest } = props;
    const theme = React.useMemo(() => {
        try {
            return themeJSON ? createTheme(JSON.parse(themeJSON) as IPartialTheme) : undefined;
        } catch (ex) {
            /* istanbul ignore next */
            console.error('Cannot parse theme', ex);
        }
    }, [themeJSON]);

    return (
        <ThemeProvider theme={theme}>
            <GaugeChart {...rest} />
        </ThemeProvider>
    );
});
CanvasGaugeChart.displayName = 'CanvasGaugeChart';
