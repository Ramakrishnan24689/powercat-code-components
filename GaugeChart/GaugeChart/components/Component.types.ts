export const enum ItemColumns {
    Legend = 'ItemLegend',
    Key = 'ItemKey',
    Count = 'ItemCount',
    Color = 'ItemColor',
}
