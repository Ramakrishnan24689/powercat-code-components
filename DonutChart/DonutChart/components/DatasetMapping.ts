import { DataVizPalette, IChartDataPoint, getColorFromToken } from '@fluentui/react-charting';
import { ItemColumns } from '../ManifestConstants';

export const _colors = [
    [DataVizPalette.color1, DataVizPalette.color2, DataVizPalette.color3, DataVizPalette.color4],
    [DataVizPalette.color5, DataVizPalette.color6, DataVizPalette.color7, DataVizPalette.color8],
    [
        DataVizPalette.color9,
        DataVizPalette.color10,
        DataVizPalette.color11,
        DataVizPalette.color12,
        DataVizPalette.color13,
    ],
    [
        DataVizPalette.color14,
        DataVizPalette.color15,
        DataVizPalette.color16,
        DataVizPalette.color17,
        DataVizPalette.color18,
    ],
    [
        DataVizPalette.color19,
        DataVizPalette.color20,
        DataVizPalette.color21,
        DataVizPalette.color22,
        DataVizPalette.color23,
        DataVizPalette.color24,
        DataVizPalette.color25,
    ],
    [
        DataVizPalette.color26,
        DataVizPalette.color27,
        DataVizPalette.color28,
        DataVizPalette.color29,
        DataVizPalette.color30,
    ],
    [
        DataVizPalette.color31,
        DataVizPalette.color32,
        DataVizPalette.color33,
        DataVizPalette.color34,
        DataVizPalette.color35,
    ],
    [
        DataVizPalette.color36,
        DataVizPalette.color37,
        DataVizPalette.color38,
        DataVizPalette.color39,
        DataVizPalette.color40,
    ],
];

export function _randomColor(index: number): string {
    const colorArray = _colors[index % _colors.length];
    return getColorFromToken(colorArray[Math.floor(Math.random() * colorArray.length)]);
}

function getKeyWithSuffix(keyIndex: Record<string, number>, key: string): string {
    if (keyIndex[key] !== undefined) {
        keyIndex[key]++;
        return `${key}_${keyIndex[key]}`;
    } else {
        keyIndex[key] = 1;
        return key;
    }
}

export function getChartDataFromDataset(
    dataset: ComponentFramework.PropertyTypes.DataSet,
    customColor: boolean,
): IChartDataPoint[] {
    if (dataset.error || dataset.paging.totalResultCount === undefined) {
        return Array.from({ length: 3 }, (_, i) => getDummyAction(`Action${i + 1}`, i));
    }
    const keyIndex: Record<string, number> = {};
    return dataset.sortedRecordIds.map((id, index) => {
        const record = dataset.records[id];
        const key = getKeyWithSuffix(keyIndex, record.getValue(ItemColumns.Key) as string);
        return {
            id: record.getRecordId(),
            key: key,
            legend: record.getValue(ItemColumns.Legend) as string,
            color: customColor ? getColorFromToken(record.getValue(ItemColumns.Color) as string) : _randomColor(index),
            data: record.getValue(ItemColumns.Value) as number,
        } as IChartDataPoint;
    });
}

function getDummyAction(key: string, index: number): IChartDataPoint {
    return {
        id: key,
        key: key,
        legend: key,
        color: _randomColor(index),
        data: 30,
    } as IChartDataPoint;
}
