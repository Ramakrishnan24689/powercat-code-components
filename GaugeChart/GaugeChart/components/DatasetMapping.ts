import { DataVizPalette, IGaugeChartSegment, getColorFromToken } from '@fluentui/react-charting';
import { ItemColumns } from '../ManifestConstants';

export const _colors = [
    [DataVizPalette.color1, DataVizPalette.color2, DataVizPalette.color3, DataVizPalette.color4],
    [DataVizPalette.color5, DataVizPalette.color6, DataVizPalette.color7, DataVizPalette.color8],
    [
        DataVizPalette.color9,
        DataVizPalette.color10,
        DataVizPalette.color11,
        DataVizPalette.color12,
        DataVizPalette.color13,
    ],
    [
        DataVizPalette.color14,
        DataVizPalette.color15,
        DataVizPalette.color16,
        DataVizPalette.color17,
        DataVizPalette.color18,
    ],
    [
        DataVizPalette.color19,
        DataVizPalette.color20,
        DataVizPalette.color21,
        DataVizPalette.color22,
        DataVizPalette.color23,
        DataVizPalette.color24,
        DataVizPalette.color25,
    ],
    [
        DataVizPalette.color26,
        DataVizPalette.color27,
        DataVizPalette.color28,
        DataVizPalette.color29,
        DataVizPalette.color30,
    ],
    [
        DataVizPalette.color31,
        DataVizPalette.color32,
        DataVizPalette.color33,
        DataVizPalette.color34,
        DataVizPalette.color35,
    ],
    [
        DataVizPalette.color36,
        DataVizPalette.color37,
        DataVizPalette.color38,
        DataVizPalette.color39,
        DataVizPalette.color40,
    ],
];

export function _randomColor(index: number): string {
    return getColorFromToken(_colors[index][Math.floor(Math.random() * _colors[index].length)]);
}

export function getChartDataFromDataset(
    dataset: ComponentFramework.PropertyTypes.DataSet,
    customColor: boolean,
): IGaugeChartSegment[] {
    if (dataset.error || dataset.paging.totalResultCount === undefined) {
        // Dataset is not defined so return dummy items
        return [getDummyAction('First', 1), getDummyAction('Second', 2), getDummyAction('Third', 3)];
    }
    const keyIndex: Record<string, number> = {};
    if (customColor) {
        return dataset.sortedRecordIds.map((id, index) => {
            const record = dataset.records[id];

            // Prevent duplicate keys by appending the duplicate index
            let key = record.getValue(ItemColumns.Key) as string;
            if (keyIndex[key] !== undefined) {
                keyIndex[key]++;
                key += `_${keyIndex[key]}`;
            } else keyIndex[key] = 1;
            return {
                id: record.getRecordId(),
                key: key,
                legend: record.getValue(ItemColumns.Legend) as string,
                size: record.getValue(ItemColumns.Size) as number,
                color: _randomColor(index),
            } as IGaugeChartSegment;
        });
    }
    return dataset.sortedRecordIds.map((id, index) => {
        const record = dataset.records[id];

        // Prevent duplicate keys by appending the duplicate index
        let key = record.getValue(ItemColumns.Key) as string;
        if (keyIndex[key] !== undefined) {
            keyIndex[key]++;
            key += `_${keyIndex[key]}`;
        } else keyIndex[key] = 1;
        return {
            id: record.getRecordId(),
            key: key,
            legend: record.getValue(ItemColumns.Legend) as string,
            size: record.getValue(ItemColumns.Size) as number,
            color: _randomColor(index),
        } as IGaugeChartSegment;
    });
}

function getDummyAction(key: string, index: number): IGaugeChartSegment {
    return {
        id: key,
        key: key,
        legend: key,
        color: _randomColor(index),
        size: 30,
    } as IGaugeChartSegment;
}
