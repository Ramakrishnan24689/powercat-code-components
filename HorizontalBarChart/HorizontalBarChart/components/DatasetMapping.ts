import { DataVizPalette, IChartProps, IHorizontalDataPoint, getColorFromToken } from '@fluentui/react-charting';
import { ItemColumns } from '../ManifestConstants';

export const _colors = [
    [DataVizPalette.color1, DataVizPalette.color2, DataVizPalette.color3, DataVizPalette.color4],
    [DataVizPalette.color5, DataVizPalette.color6, DataVizPalette.color7, DataVizPalette.color8],
    [
        DataVizPalette.color9,
        DataVizPalette.color10,
        DataVizPalette.color11,
        DataVizPalette.color12,
        DataVizPalette.color13,
    ],
    [
        DataVizPalette.color14,
        DataVizPalette.color15,
        DataVizPalette.color16,
        DataVizPalette.color17,
        DataVizPalette.color18,
    ],
    [
        DataVizPalette.color19,
        DataVizPalette.color20,
        DataVizPalette.color21,
        DataVizPalette.color22,
        DataVizPalette.color23,
        DataVizPalette.color24,
        DataVizPalette.color25,
    ],
    [
        DataVizPalette.color26,
        DataVizPalette.color27,
        DataVizPalette.color28,
        DataVizPalette.color29,
        DataVizPalette.color30,
    ],
    [
        DataVizPalette.color31,
        DataVizPalette.color32,
        DataVizPalette.color33,
        DataVizPalette.color34,
        DataVizPalette.color35,
    ],
    [
        DataVizPalette.color36,
        DataVizPalette.color37,
        DataVizPalette.color38,
        DataVizPalette.color39,
        DataVizPalette.color40,
    ],
];

export function _randomColor(index: number): string {
    const colorGroup = _colors[index % _colors.length];
    return getColorFromToken(colorGroup[Math.floor(Math.random() * colorGroup.length)]);
}

function createChartProps(
    record: ComponentFramework.PropertyHelper.DataSetApi.EntityRecord,
    key: string,
    index: number,
    ariaLabel: string | undefined,
    customColor: boolean,
): IChartProps {
    return {
        id: record.getRecordId(),
        key,
        chartTitle: record.getValue(ItemColumns.Title) as string,
        chartTitleAccessibilityData: ariaLabel,
        chartData: [
            {
                horizontalBarChartdata: {
                    x: record.getValue(ItemColumns.Value) as number,
                    y: record.getValue(ItemColumns.TotalValue) as number,
                } as IHorizontalDataPoint,
                legend: record.getValue(ItemColumns.Legend) as string,
                xAxisCalloutData: record.getValue(ItemColumns.XPopover) as string,
                yAxisCalloutData: record.getValue(ItemColumns.YPopover) as string,
                color: customColor
                    ? getColorFromToken(record.getValue(ItemColumns.Color) as string)
                    : _randomColor(index),
            },
        ],
    } as IChartProps;
}

export function getChartDataFromDataset(
    dataset: ComponentFramework.PropertyTypes.DataSet,
    customColor: boolean,
    ariaLabel: string | undefined,
): IChartProps[] {
    if (dataset.error || dataset.paging.totalResultCount === undefined) {
        return Array.from({ length: 3 }, (_, index) => getDummyAction(`Action ${index + 1}`, index + 1));
    }
    const keyIndex: Record<string, number> = {};
    return dataset.sortedRecordIds.map((id, index) => {
        const record = dataset.records[id];
        const key = record.getValue(ItemColumns.Key) as string;
        keyIndex[key] = (keyIndex[key] || 0) + 1;
        return createChartProps(
            record,
            key + (keyIndex[key] > 1 ? `_${keyIndex[key] - 1}` : ''),
            index,
            ariaLabel,
            customColor,
        );
    });
}

function getDummyAction(key: string, index: number): IChartProps {
    return {
        id: key,
        key,
        chartTitle: key,
        chartData: [
            {
                horizontalBarChartdata: {
                    x: index * 10,
                    y: 50,
                } as IHorizontalDataPoint,
                legend: key,
                color: _randomColor(index),
            },
        ],
    } as IChartProps;
}
