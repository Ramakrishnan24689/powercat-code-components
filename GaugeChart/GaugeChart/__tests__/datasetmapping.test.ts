import { ItemColumns } from '../ManifestConstants';
import { MockDataSet, MockEntityRecord } from '../__mocks__/mock-datasets';
import { getChartDataFromDataset } from '../components/DatasetMapping';

describe('DatasetMapping', () => {
    beforeEach(() => jest.clearAllMocks());

    it('returns correct props when custom color is false', () => {
        const items = [
            new MockEntityRecord('1', {
                [ItemColumns.Key]: 'item1',
                [ItemColumns.Legend]: 'Item 1',
                [ItemColumns.Size]: 30,
                [ItemColumns.Color]: 'blue',
            }),
            new MockEntityRecord('1', {
                [ItemColumns.Key]: 'item1',
                [ItemColumns.Legend]: 'Item 1',
                [ItemColumns.Size]: 20,
                [ItemColumns.Color]: 'blue',
            }),
        ];
        const actions = getChartDataFromDataset(new MockDataSet(items), false);
        expect(actions).toMatchSnapshot();
    });

    it('returns correct props when custom color is true ', () => {
        const items = [
            new MockEntityRecord('1', {
                [ItemColumns.Key]: 'item1',
                [ItemColumns.Legend]: 'Item 1',
                [ItemColumns.Size]: 30,
                [ItemColumns.Color]: 'blue',
            }),
            new MockEntityRecord('1', {
                [ItemColumns.Key]: 'item1',
                [ItemColumns.Legend]: 'Item 1',
                [ItemColumns.Size]: 20,
                [ItemColumns.Color]: 'blue',
            }),
        ];
        const actions = getChartDataFromDataset(new MockDataSet(items), true);
        expect(actions).toMatchSnapshot();
    });
});
