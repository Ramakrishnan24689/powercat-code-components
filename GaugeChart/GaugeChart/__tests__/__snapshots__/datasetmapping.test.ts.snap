// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DatasetMapping returns correct props when custom color is false 1`] = `
Array [
  Object {
    "color": "#e3008c",
    "id": "1",
    "key": "item1",
    "legend": "Item 1",
    "size": 20,
  },
  Object {
    "color": "#57811b",
    "id": "1",
    "key": "item1_2",
    "legend": "Item 1",
    "size": 20,
  },
]
`;

exports[`DatasetMapping returns correct props when custom color is true  1`] = `
Array [
  Object {
    "color": "#e3008c",
    "id": "1",
    "key": "item1",
    "legend": "Item 1",
    "size": 20,
  },
  Object {
    "color": "#13a10e",
    "id": "1",
    "key": "item1_2",
    "legend": "Item 1",
    "size": 20,
  },
]
`;
